### ISSUE TITLE
VoidValue proposal

### ISSUE THREAD

**Issue by vsbogd** [opencog/atomspace#2627]
`GroundedSchemeNode` functions are required to return some result and they cannot return `Handle::INVALID` because execution code calls methods on returned value: https://github.com/opencog/atomspace/blob/36a98d775e1ca294eedb590135c61a47f1a945ef/opencog/atoms/execution/Instantiator.cc#L678-L680

I propose introducing `VoidValue` to solve the issue. The value which has only instance and separate VOID_TYPE

**Comment by linas** [comment]
What is the actual problem that you are trying to solve?

**Comment by vsbogd** [comment]
I am trying to execute GroundedSchemaNode and it fails with segfault in Instantiator.cc

**Comment by linas** [comment]
The meta-issue is that having allowing `Handle::UNDEFINED` aka nullptr handles seems to have been a design mistake, so I am trying to get rid of every location where they might occur.  Do you have a stack trace?  (Most recently, `MapLink` was returning `Handle::UNDEFINED` but that was fixed just 3-4 days ago.)

**Comment by vsbogd** [comment]
Following script fails:
```py
import unittest
from opencog.exec import execute_atom
from opencog.type_constructors import *

class VoidFuncTest(unittest.TestCase):

    atomspace = AtomSpace()

    def setUp(self):
        set_default_atomspace(self.atomspace)
        InheritanceLink(ConceptNode("Frog"),       ConceptNode("animal"))

    def test_void_func(self):
        bind = BindLink(VariableNode("x"),
                 AndLink(PresentLink(VariableNode("x"))),
                 ExecutionOutputLink(
                     GroundedSchemaNode("py: void_func"),
                     ListLink(VariableNode("x"))))
        result = execute_atom(self.atomspace, bind)

def void_func(atom):
    print("function called")
    pass

if __name__ == "__main__":
    unittest.main()
```

Failure description:
```
======================================================================
ERROR: test_void_func (__main__.VoidFuncTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "../atomspace/tests/cython/bindlink/test_void_func.py", line 19, in test_void_func
    result = execute_atom(self.atomspace, bind)
  File "exec.pyx", line 14, in opencog.exec.execute_atom
AttributeError: 'NoneType' object has no attribute 'handle_ptr'

----------------------------------------------------------------------
```

**Comment by linas** [comment]
Umm, don't do that? User error? The obvious fix is to alter `ExecutionOutput` to throw an error if it did not get something valid (which we should do anyway, just in case the user returns assorted unexpected garbage...)

I'm assuming that you really want to do things this way for some reason ... but I'm pushing back, because after long hard experience, it seems that allowing null/empty handles is a design mistake (There are maybe a dozen different design mistakes, all reinforcing one-another, untangling them and fixing them is hard work, its like weeds with really strong roots. I don't really want to let in one more thing unless its somehow very obviously makes something better.)

**Comment by vsbogd** [comment]
OK, the question is what should I return from function? Are you suggesting get any random value and return?

**Comment by linas** [comment]
>  what should I return from function? 

You were passed `atom` -- you can safely return that.

FWIW .. at least some of the code does check  for unexpected return values, but not the python bindings.  I'm writing unit tests now.

**Comment by vsbogd** [comment]
If my function is pure side effect and have no parameters, what should I return then? I can just construct new TruthValue and return it but why should I do that? Previously I returned Handle::UNDEFINED, but now it fails with segfault:
```
Program received signal SIGSEGV, Segmentation fault.
opencog::Instantiator::execute (this=0x7fffffffe208, expr=..., silent=<optimized out>) at /home/vital/projects/opencog/atomspace/opencog/atoms/execution/Instantiator.cc:679
679                     if (vp->is_atom())
```

**Comment by linas** [comment]
Well, if you change `if (vp->is_atom())` to `if (vp and vp->is_atom())` then you should have gotten an exception
```
  File "exec.pyx", line 14, in opencog.exec.execute_atom
AttributeError: 'NoneType' object has no attribute 'handle_ptr_object'
```
or something similar. At least, on reading the code, that is what python seems to currently be doing, if one returns junk.  I'm crawling through the cython code now -- there were lots of checks for bad return values, yet your code somehow managed to squeeze past all of them, un-noticed. I'm trying to figure that out now.

**Comment by linas** [comment]
So  #2629 makes all error checking uniform.  Before this pull req, I'm confused about
how your test case failed to trigger a throw. There is a check here:

https://github.com/opencog/atomspace/blob/f908901010e45d5e80301f2767feb7ec22a0c32b/opencog/cython/PythonEval.cc#L996-L999

Yet apparently in python, `pass` actually returns `None` which is a valid cython object, so that check didn't stop it.  Next, there's this check:

https://github.com/opencog/atomspace/blob/f908901010e45d5e80301f2767feb7ec22a0c32b/opencog/cython/PythonEval.cc#L1005-L1011

and `None` should not have an attribute called `handle_ptr` so that check should have triggered.  That should have caused this to print:
```
  File "exec.pyx", line 14, in opencog.exec.execute_atom
AttributeError: 'NoneType' object has no attribute 'handle_ptr'
```  
So I'm stumped on how you managed to not get that error.  Anyway, that error message was opaque to normal atomspace users, so I added an explicit check to see if the attribute exist,  instead of assuming that it does, and just calling it:
```
 if (0 == PyObject_HasAttrString(pyValue, "handle_ptr"))
```
that definitely does catch the error. That's more or less the "only" "real" change in pull-req #2629 everything else in there just adds unit tests, and merges the older two cut-n-paste blocks of code back into one.

**Comment by vsbogd** [comment]
Code in #2629 looks good, but still there is a question: what should I return from function which is not supposed to return value. Do you propose pick some random value and return?

**Comment by linas** [comment]
Well, you still haven't explained why you need this. 

It makes no sense to call a GSN that takes no arguments and returns no values. Its just a highly inefficient waste of CPU cycles. You can do it much more efficiently just by calling `for(i=0; i<n; i++) f();` which will have the same effect, but without the cpu-wastage.

It also opens the door to a separate set of questions: if we're going to automatically convert a python `None` to `VoidValue`, then shouldn't we automatically convert python `String` to `StringValue`, python `List` to `LinkValue` and so on? This is much more interesting. The pros and cons seem to be:

* It adds more CPU overhead, which is pointless if no one cares about the results.
* Code that does care about the results would have to convert from `BindLink` to  `QueryLink` -- because `BindLink` returns a `Set` of Atoms, while `QueryLink` return a `LinkValue` of Values (but is otherwise exactly the same thing). 
* Code that does care about this can easily do this by hand ...  Its not hard.... e.g. instead of having the GSN do a `return "foobar"` and hope for an automatic conversion, you instead can say `return StringValue("foobar")` - so zero lines of code, extra.
 
The biggest argument against is that using GSN's is mostly useless and pointless, they offer very little or no advantage, and just waste CPU time. That is, instead of calling `Bind(...stuff... GSN(x,y,z))` instead just call `Get(...stuff...);  for(i=0, i<n; i++) f(x,y,z)` that is, let the GetLink find the stuff you need, and then iterate over what Get returns. You get the same effect, but without the overhead of entering-leaving python for each GSN call (and without the overhead of a `VoidValue` conversion). 

I mean .. `VoidValue` is particularly pointless with `BindLink`, as all that happens is that the VoidValue ctor runs, malloc memory, then the c++ smart-pointer runs, mallocs memory, then `BindLink` ignores it cause its not an atom, then the c++ smart-pointer runs and decrements the count, then the  c++ smart-pointer self-destructs, then the finalizer for VoidValue runs (cause the count is now zero, and there are no weak pointers ...yes more cpu wasted on weak pointers) and finally, the VoidValue dtor runs. What was the point of all of that malloc'ing and freeing? Its maybe a a few hundred cpu instructions, and maybe a dozen cache-misses ... to do nothing at all...

**Comment by linas** [comment]
BTW, the overhead of using GSN's is non-trivial. I have not measured recently, but it used to be capped at about 10K/second or 20K/second, for a no-op GSN. so that is 100 micro-seconds to 200 microseconds per call, ... on a 2GHz machine, that is about 2K cpu-cycles per microsecond, or 200K to 400K cpu-cycles per no-op GSN.  On a raspberry-pi doing robot control, its even more scandalous. 

The moral of the story is that Atomese is perfect for knowledge representation, and performing complex queries.   Atomese shouldn't (currently?) be used as a generic compute platform, and the GSN/GPN mechanism is supposed to be an emergency use-as-a-last-resort system.  Much much more interesting is the question of "what are you doing with the GSN/GPN, and can we invent an Atom that does the same thing, directly?" -- that is the very interesting question that can be pursued.

Well, I guess that there is also a different, futuristic direction to go in ... "can we compile Atomese down to python bytecode?" (or scheme byte-code, etc.)  That could be very interesting for performance. For example, `(SetValue (Predicate "counter") (Concept "thing") (Plus (Number 1) (GetValue  (Predicate "counter") (Concept "thing"))))` is a complicated way of incrementing a counter on a "thing". We can either -- invent `(IncrementLink (Predicate "counter") (Concept "thing") (Number 1))` or we can "compile" the longer thing into bytecode, and run the bytecode.  But this is a ... long and difficult experiment. Interesting, but a sea-change.

**Comment by vsbogd** [comment]
In my example `GSN` does something in a real word, for example switch the light on. So using pure atomese for doing it is likely not applicable.

**Comment by bgoertzel** [comment]
In the example of turning the light on, I guess the function could
meaningfully return a "success" signal meaning "yeah I sent the signal
to turn the light on."   However, it could also be the case that a
success signal takes too long to wait for and would have to come via a
separate message sent back to Atomspace.  In which case indeed you
need a VoidValue or something semantically equivalent.

On Mon, May 25, 2020 at 12:52 AM Vitaly Bogdanov
<notifications@github.com> wrote:
>
> In my example GSN does something in a real word, for example switch the light on. So using pure atomese for doing it is likely not applicable.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.



-- 
Ben Goertzel, PhD
http://goertzel.org

“The only people for me are the mad ones, the ones who are mad to
live, mad to talk, mad to be saved, desirous of everything at the same
time, the ones who never yawn or say a commonplace thing, but burn,
burn, burn like fabulous yellow roman candles exploding like spiders
across the stars.” -- Jack Kerouac

**Comment by linas** [comment]
You can make this work today by returning, for example, a truth value, or a float value of length zero, or anything.

As noted earlier, creating a special object that expends thousands of CPU cycles to do absolutely nothing at all is not wise system design.  Other programming languages have a "void" return value, because they can handle it in zero cycles.  If someone can figure out how to create a zero-cycle "void" for atomese, that would be great, and I have no objections to that. 

BTW, returning null pointers is not really a valid alternative -- null points require if-checks for null to be sprinkled everywhere, which hurts in two ways: (1) it costs extra cpu cycles, and (2) code will crash with null pointer dereferences during trade shows, investor demos, and any time something important and urgent comes up. Best to avoid them.

To repeat: if you have a zero-cycle void-value proposal, great! Otherwise, not.

**Comment by linas** [comment]
Maybe it ifs a static global singleton or something like that. ...

**Comment by vsbogd** [comment]
Yes, the original idea is to make VoidValue constructor private and create only instance say VoidValue::INSTANCE. The difference with Handle::UNDEFINED is VoidValue::INSTANCE don't require NULL checks. And it costs nothing because it is static and created only once.

**Comment by linas** [comment]
> the original idea 

OK, yes, that seems reasonable. Make the factory return the singleton instance, too, and I'll merge the pull req.

**Comment by einalex** [comment]
> For example, `(SetValue (Predicate "counter") (Concept "thing") (Plus (Number 1) (GetValue (Predicate "counter") (Concept "thing"))))` is a complicated way of incrementing a counter on a "thing". We can either -- invent `(IncrementLink (Predicate "counter") (Concept "thing") (Number 1))`

I was writing one of these just yesterday. I think an IncrementLink (and maybe syntactic sure DecrementLink) would be great.

**Comment by vsbogd** [comment]
Implemented in #2667

**Comment by linas** [comment]
@einalex can you open a new issue explicitly requesting this? It seems like decrement is same as increment minus 1 unless you want it to stop at 0. 

BTW, be aware that, under the covers, `NumberNode` and `FloatValue` are actually vectors, so add and subtract are actually vector add and subtract... the reason for vectors is a general plan  to provide better support for integrating with neural-net architectures.  (e.g. to have some link that launches a tensorflow job on some gpu's or whatever).  (well, my understanding is that @vsbogd already has code that does this, but instead of creating a new git-repo, he jammed that code into a clone of the atomspace repo, and has been paying the price for this ever since. So, @vsbogd, you should really, really split out that code, so that it lives in it's own repo, instead of tangling it into the atomspace... that would make integration, unit testing, maintenance a lot easier.)

**Comment by vsbogd** [comment]
> split out that code, so that it lives in it's own repo, instead of tangling it into the atomspace... that would make integration, unit testing, maintenance a lot easier.

Unfortunately, practice shows that it is not easier. In last merge for example this code was broken by changes in atomspace. So while this code is not part of the CI it is not easier.

**Comment by linas** [comment]
> practice shows that it is not easier. 

I don't understand ... if you created a repo, say, https://github.com/opencog/visual-answer or https://github.com/singnet/visual-answer then it could be a part of CI. This is analogous to https://github.com/opencog/attention and to https://github.com/opencog/spacetime and https://github.com/opencog/agi-bio all of which add new atom types that do new and strange things. Attention even adds new values -- the `AttentionValue` which has it's own API and behavior, and 8 more new atom types that know how to do stuff with it ...

**Comment by linas** [comment]
Closing #2667 has been merged
